src/invidious/helpers/signatures.cr

```diff
@@ -70,4 +75,36 @@ struct DecryptFunction
 
     return "&#{sp}=#{sig.join("")}"
   end
+
+  def overwrite_n(id : String, fmt : Hash(String, JSON::Any))
+    uri = URI.parse(fmt["url"].as_s)
+    params = HTTP::Params.parse(uri.query.not_nil!)
+    return fmt["url"].as_s unless params["n"]?
+    n = params["n"]
+
+    function_name = "(cache #1)"
+    if (@last_decrypted_n[:id] == id && @last_decrypted_n[:n] == n)
+      dec_n = @last_decrypted_n[:dec_n]
+    else
+      document = YT_POOL.client &.get("/watch?v=#{id}&gl=US&hl=en").body
+      playerjs = document.match(/src="(?<url>\/s\/player\/[^\/]+\/player_ias[^\/]+\/en_US\/base.js)"/).not_nil!["url"]
+      function_name = "(cache #2)"
+      function_body = @playerjs_decrypt_n_function[playerjs]?
+      if (!function_body)
+        player = YT_POOL.client &.get(playerjs).body
+        function_name = player.match(/a\.get\("n"\)\)&&\(b=(?<nfunc>[a-zA-Z0-9]+)\(b\)/m).not_nil!["nfunc"]
+        function_body = player.match(/^#{Regex.escape(function_name)}=(?<body>function\(\w\)\{.*?"enhanced_except_[^\}]+\}[^\}]+\})/m).not_nil!["body"]
+        @playerjs_decrypt_n_function[playerjs] = function_body
+      end
+      rt = Duktape::Runtime.new do |sbx|
```

Comment:
You should probably use a sandbox, instead of a runtime, as this javascript is what we can consider "untrusted"...

Reply:
Doesn't the runtime use a sandbox by default? (I'm not very familiar with crystal...)
https://github.com/jessedoyle/duktape.cr/blob/master/src/duktape/runtime.cr

Reply:
Ah, you are right. Their documentation suggests the opposite ^^